src/main/scala/com/evobootcamp/homeworks/db/Doobie.scala

```diff
@@ -0,0 +1,428 @@
+        year.map(y => fr"year = $y")
+    ))
+    .query[Book]
+    .to[List]
+  }
+
+  def findBook(id: UUID): ConnectionIO[Either[DbError, BookWithAuthor]] = {
+    sql"""SELECT b.id, b.title, b.year, b.genre, a.id, a.name, a.birthday
+         | FROM books b
+         | JOIN authors a ON a.id = b.author
+         | WHERE b.id = $id"""
+      .stripMargin
+      .query[BookWithAuthor]
+      .option
+      .map {
+        case Some(b) => Right(b)
+        case None    => Left(Errors.BookNotFound)
+      }
+  }
+
+  def deleteBook(id: UUID): ConnectionIO[Either[DbError, Unit]] = {
+    sql"""DELETE FROM books WHERE id = $id"""
+      .update
+      .run
+      .map(res => if (res > 0) Right() else Left(Errors.BookNotFound))
+  }
+
+  def runQuery[F[_] : ContextShift : Async, A](query: ConnectionIO[A]): F[A] = {
+    Connection.pooled[F].use { cn =>
+      for {
+        result <- query.transact(cn)
+      } yield result
+    }
+  }
+}
+
+object Doobie extends IOApp {
+  import Models._
+  import Db._
+  import Db.Errors._
+  import Protocol._
+
+  import io.circe.generic.auto._
+  import org.http4s.circe.CirceEntityCodec._
+
+  private case class BadResponse(message: String)
+
+  private def getErrorMessage(e: DbError): String = e match {
+    case DatabaseError  => "Database query error"
+    case AuthorNotFound => "Author is not found"
+    case BookNotFound   => "Book is not found"
+    case _ => "Internal error"
+  }
+
+  private def renderResult[A](r: Either[DbError, A])(implicit ee: EntityEncoder[IO, A]): IO[Response[IO]] = {
+    r match {
+      case Left(e) => e match {
+        case BookNotFound => NotFound(BadResponse(getErrorMessage(e)))
+        case e => BadRequest(BadResponse(getErrorMessage(e)))
+      }
+      case Right(v) => Ok(v)
+    }
+  }
+
+  private object CRUDController {
+    def update[A, B](req: Request[IO], query: A => ConnectionIO[Either[DbError, B]])
+    (implicit ed: EntityDecoder[IO, A], ee: EntityEncoder[IO, B]): IO[Response[IO]] = {
+      req.as[A].flatMap { data =>
+        for {
+          item <- runQuery[IO, Either[DbError, B]](query(data))
+          result <- renderResult[B](item)
+        } yield result
+      }
+    }
+
+    def delete(query: ConnectionIO[Either[DbError, Unit]]): IO[Response[IO]] = {
+      for {
+        res <- runQuery[IO, Either[DbError, Unit]](query)
+        result <- renderResult[Unit](res)
+      } yield result
+    }
+
+    def findOne[B](query: ConnectionIO[Either[DbError, B]])
+    (implicit ee: EntityEncoder[IO, B]): IO[Response[IO]] = {
+      for {
+        item <- runQuery[IO, Either[DbError, B]](query)
+        result <- renderResult[B](item)
+      } yield result
+    }
+
+    def findAll[B](query: ConnectionIO[List[B]])
+               (implicit ee: EntityEncoder[IO, List[B]]): IO[Response[IO]] = {
+      for {
+        items <- runQuery[IO, List[B]](query).handleErrorWith(e => {
+          println(e)
+          IO.pure(List())
```

Comment:
`List()` allocates new empty list. I would even consider it a side effect to some extent and replace `IO.pure` with simple `IO(..)` :). It's preferable to use `List.empty` which is effectively `Nil` not to create new instances of `List`.